osquery/filesystem/tests/filesystem.cpp

```diff
@@ -951,4 +947,57 @@ TEST_F(FilesystemTests, test_directory_listing_with_legacy_logic) {
   deleteDirectoryContent(test_root_work_dirs);
 }
 
+TEST_F(FilesystemTests, test_directory_listing_with_file_symlink) {
+  // This test verifies that a file symlink is not mistaken for a directory.
+  const fs::path test_root_dir = fs::temp_directory_path() / genRandomName();
+  ASSERT_TRUE(fs::create_directory(test_root_dir));
+
+  fs::ofstream test_file(test_root_dir / "test_file.txt");
+  test_file.close();
+
+  // Create symlink
+  try {
+    fs::create_symlink(test_root_dir / "test_file.txt", test_root_dir / "link");
+  } catch (const fs::filesystem_error& e) {
+    FAIL() << "Error creating symlink: " << e.what();
+  }
+
+  std::vector<std::string> found_directories;
+  ASSERT_TRUE(
+      listDirectoriesInDirectory(test_root_dir, found_directories, false));
+
+  ASSERT_TRUE(found_directories.empty());
+
+  deleteDirectoryContent(test_root_dir);
+}
+
+TEST_F(FilesystemTests, test_directory_listing_with_bad_symlinks) {
+  // This test verifies that bad symlinks are not mistaken for a directory.
+  const fs::path test_root_dir = fs::temp_directory_path() / genRandomName();
+  ASSERT_TRUE(fs::create_directory(test_root_dir));
+
+  // Create symlink that points to itself.
+  try {
+    fs::create_symlink(test_root_dir / "link", test_root_dir / "link");
+  } catch (const fs::filesystem_error& e) {
+    FAIL() << "Error creating symlink: " << e.what();
+  }
+
+  // Create symlink that points to non-existent file.
+  try {
+    fs::create_symlink(test_root_dir / "not_exists.txt",
+                       test_root_dir / "link2");
+  } catch (const fs::filesystem_error& e) {
+    FAIL() << "Error creating symlink: " << e.what();
+  }
+
+  std::vector<std::string> found_directories;
+  ASSERT_TRUE(
+      listDirectoriesInDirectory(test_root_dir, found_directories, false));
```

Comment:
Same as the other comment, let's test with `recursive=true`.

Reply:
Added